pkg/controller/mpi_job_controller.go

```diff
@@ -118,6 +118,9 @@ const (
 
 	openMPISlotsEnv  = "OMPI_MCA_orte_set_default_slots"
 	intelMPISlotsEnv = "I_MPI_PERHOST"
+
+	// Annotation instructing Kueue to ignore a workload for scheduling.
+	KueueIgnoreAnnotation = "kueue.x-k8s.io/ignore"
```

Comment:
Can we merge this change into the master branch after the release of kueue v0.3.0, including https://github.com/kubernetes-sigs/kueue/pull/558? 

Also, it would be good to import this annotation from kueue instead of hard coding.


cc: @alculquicondor

Reply:
> Can we merge this change into the master branch after the release of kueue v0.3.0, including [kubernetes-sigs/kueue#558](https://github.com/kubernetes-sigs/kueue/pull/558)?

We could wait, but I don't see a clear benefit. Once the Kueue PR is merged the rename of the annotation will be unlikely. Also, adding the annotation in case running with Kueue 0.2.0 will be harmless and the Kueue will simply ignore it. One advantage of merging early is that we will be able to continue the remaining Kueue work (the operator to suspend / unsuspend the MPIJob) using master of the mpi-operator, and not a branch.

> Also, it would be good to import this annotation from kueue instead of hard coding.

I agree, however I don't think we want MPI operator to depend on the entire Kueue, so we would need to first extract a small kueue-api library with the API for integrations. I guess it could be providing the interfaces for integration similar to those suggested in the KEP: https://github.com/kubernetes-sigs/kueue/pull/537. However, I think it would be good to first do one or two example integrations to see how the integration looks like and then extract the library, but possibly for 0.4.0.

Reply:
> One advantage of merging early is that we will be able to continue the remaining Kueue work (the operator to suspend / unsuspend the MPIJob) using master of the mpi-operator, and not a branch.

IIRC, we are planning to add workload-controller for MPIJob (like https://github.com/kubernetes-sigs/kueue/tree/main/pkg/controller/workload/job) to mpi-operator repository. So it seems that we will be able to continue the remaining Kueue work without merging this PR. WDYT?

ref: https://github.com/kubeflow/mpi-operator/issues/505

Reply:
> I guess it could be providing the interfaces for integration similar to those suggested in the KEP: https://github.com/kubernetes-sigs/kueue/pull/537. However, I think it would be good to first do one or two example integrations to see how the integration looks like and then extract the library, but possibly for 0.4.0.

It sounds good to me.

Reply:
> IIRC, we are planning to add workload-controller for MPIJob (like https://github.com/kubernetes-sigs/kueue/tree/main/pkg/controller/workload/job) to mpi-operator repository. So it seems that we will be able to continue the remaining Kueue work without merging this PR. WDYT?



> IIRC, we are planning to add workload-controller for MPIJob (like https://github.com/kubernetes-sigs/kueue/tree/main/pkg/controller/workload/job) to mpi-operator repository. So it seems that we will be able to continue the remaining Kueue work without merging this PR. WDYT?
> 
> ref: #505

Technically it could be either in mpi-operator or in Kueue, I don't have a strong opinion where. The argument for putting it into Kueue is that then we have a clear view of how the integrations look like and so we may have a better idea of how the common library should look like. 

Still, I don't see a clear benefit in holding the annotation task. It should also not be influencing the decision about the repo. 

cc: @mwielgus.

Reply:
> Technically it could be either in mpi-operator or in Kueue, I don't have a strong opinion where. The argument for putting it into Kueue is that then we have a clear view of how the integrations look like and so we may have a better idea of how the common library should look like.

I see. It looks like you have a point. Let me know what others think. cc: @alculquicondor @ahg-g @kerthcet

However, If we add a workload-controller for MPIJob to the kueue repository, it might be better to implement the logic to propagate `.metadata.annotations` of MPIJob to Launcher and Worker instead of hard coding `kueue.x-k8s.io/ignore` (this means the mpi-operator propagates annotations of MPIJob to Launcher batch/v1 Job and Worker Pods). And then, a workload-controller injects `kueue.x-k8s.io/ignore` to annotations of MPIJob.

Since it would be unnecessary for users who don't use kueue.

WDYT?

Reply:
> However, If we add a workload-controller for MPIJob to the kueue repository, it might be better to implement the logic to propagate `.metadata.annotations` of MPIJob to Launcher and Worker instead of hard coding `kueue.x-k8s.io/ignore` (this means the mpi-operator propagates annotations of MPIJob to Launcher batch/v1 Job and Worker Pods). And then, a workload-controller injects `kueue.x-k8s.io/ignore` to annotations of MPIJob.

Let me check if I understand correctly. You suggest that the mpi-operator (in this repo) should propagate all annotations it got set by the workload-controller for MPIJob in Kueue down to the Launcher. However, it would mean that we need to add the `kueue.x-k8s.io/ignore` to MPIJob, and this is not what we want. We want the MPIJob to be scheduled by Kueue, not ignored. Thus we need to have the annotation only at the Launcher level, but not at the MPIJob level.

Reply:
> Let me check if I understand correctly. You suggest that the mpi-operator (in this repo) should propagate all annotations it got set by the workload-controller for MPIJob in Kueue down to the Launcher. 

Yes, that's right.  

> However, it would mean that we need to add the kueue.x-k8s.io/ignore to MPIJob, and this is not what we want. We want the MPIJob to be scheduled by Kueue, not ignored. Thus we need to have the annotation only at the Launcher level, but not at the MPIJob level.

As you say, it might be weird that Workers have `kueue.x-k8s.io/ignore` annotations. 
So if the kueue repo has a workload-controller for MPIJob, this PR change might be a better way.

Thanks for clarifying.

Reply:
IIUC It means no changes required in the mpi-operator, so closing the PR.

Reply:
Sounds good. I prefer the way to insert annotation by webhook on kueue side.

Reply:
Yes, being able to integrate without having to inject kueue-specific logic is a great benefit!